service/topology_coordinator.cc

```diff
@@ -440,15 +441,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                         " can't find endpoint for token {}", end));
                 }
 
-                auto ptr = _topo_sm._topology.find(raft::server_id{ep->uuid()});
-                if (!ptr) {
-                    on_internal_error(rtlogger, ::format(
-                        "make_new_cdc_generation_data: get_sharding_info:"
-                        " couldn't find node {} in topology, owner of token {}", *ep, end));
+                try {
+                    return get_sharding_info_for_host_id(*ep);
+                } catch (...) {
+                    std::throw_with_nested(std::runtime_error(::format(
+                        "raft topology: make_new_cdc_generation_data: get_sharding_info:"
+                        " can't get sharding info for node {}, owner of token {}", *ep, end)));
```

Comment:
Hm, right, that was not intentional for the existing cases (i.e. when the function is called under other circumstances than upgrade) as it uses info from topology which should be consistent.

I'd rather not trigger an internal error during upgrade - if something is wrong and the gossip state is broken it's better to throw an exception and retry rather than: 1) crash the current node 2) elect a new leader and start the topology coordinator on it 3) goto 1.

I'll try to reorganize so that it throws in the right circumstances and causes an internal error in the remaining ones.

Reply:
Lets ask @kbr-scylla if it have to be internal_error in non upgrade case and what will happen if we throw instead.

Reply:
If the information is missing in system.topology (non upgrade case), it's a bug, and bugs are marked with on_internal_error.

Reminder: on_internal_error doesn't crash in production, unless someone sets `--abort-on-internal-error` (not recommended), but it will crash in tests, which is good to catch issues early especially in test.py which doesn't detect ERRORs in logs.

Reply:
> Reminder: on_internal_error doesn't crash in production

Right, I forgot about that. In that case it makes sense to go with it in both cases.